Adds Temperature Control Units, a.k.a space heaters

Menshin opened the pull request:
<!-- Please read these guidelines before opening your PR: https://docs.spacestation14.io/en/getting-started/pr-guideline -->
<!-- The text between the arrows are comments - they will not be visible on your PR. -->

## About the PR
<!-- What did you change in this PR? -->
This adds space "heaters" (which can also cool air, so...).
Needs #25247 and closes #17181.
## Why / Balance
<!-- Why was it changed? Link any discussions or issues here. Please discuss how this would affect game balance. -->
There's a few missing atmos devices from SS13 and space heaters are one of these. Also, presently there's no real way to control temperatures outside of venting/opening a canister of pre-heat/pre-cooled air.

## Technical details
<!-- If this is a code change, summarize at high level how your new code works. This makes it easier to review. -->
Adds a `spaceHeater` component that can control a `thermoMachine` component behavior via its `Cp` (coefficient of performance).
Basically, the spaceHeater has a positive `Cp` for heating and a negative one for cooling and makes the thermomachine switch between these depending on the local mixture temperature.

Like in SS13, there's three working modes :
* Heat : begins heating the atmosphere if the temperature gets below the target temperature + the temperature tolerance
* Cold : same with cooling when above the targetTemp + tolerance
* Auto : adapts the heat exchange direction to stay as close as possible of the targetTemp

The space heater is APC powered, because it was chewing on batteries pretty fast.

I've also added a maintenance panel (useless for now, but the sprite was there, so...) and a board to construct it (currently only available on admin spawn, no idea if it should be tied to a given tech or freely available on roundstart lathes).
Localization is used too.

Also fixed some bugs with the portable scrubber while i was at it (no collision on spawn, giving possibly more glass that what was used to construct it and asserting on construction) .
Also some prototype cleaning.

**Any and all default values are purely arbitraries and roughly balanced, i'll gladly change any on enlightened suggestions.**

## Media
<!-- 
PRs which make ingame changes (adding clothing, items, new features, etc) are required to have media attached that showcase the changes.
Small fixes/refactors are exempt.
Any media may be used in SS14 progress reports, with clear credit given.

If you're unsure whether your PR will require media, ask a maintainer.

Check the box below to confirm that you have in fact seen this (put an X in the brackets, like [X]):
-->


https://github.com/space-wizards/space-station-14/assets/7117411/ef315f2b-c1d5-4108-a92d-299b957cbdf7

Updated UI with power level selector :
![image](https://github.com/space-wizards/space-station-14/assets/7117411/c510f100-b8b7-4828-bab0-b50a594b39f3)



- [X] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

## Breaking changes
<!--
List any breaking changes, including namespace, public class/method/field changes, prototype renames; and provide instructions for fixing them. This will be pasted in #codebase-changes.
-->
None i could find.

**Changelog**
<!--
Make players aware of new features and changes that could affect how they play the game by adding a Changelog entry. Please read the Changelog guidelines located at: https://docs.spacestation14.io/en/getting-started/pr-guideline#changelog
-->

- add: Space heaters are now available to heat or cool rooms.

Plykiya commented:
> presently there's no real way to control temperatures outside of venting/opening a canister of pre-heat/pre-cooled air.

The way to control temperatures that's currently done is either using a heater or a freezer and doing an in-room loop via a vent intake/output so it constantly cycles the same air in the room. Science does it often to deal with their hyper convection autolathes and whatnot.

Menshin commented:
Yeah, i was thinking of pipeless temperature control, but this works fine of course 👍

Ilya246 commented:
the stats seem weak to me, tile air generally needs __a lot__ more heat to cool or heat than pipe air

Partmedia commented on Content.Client/Atmos/UI/SpaceHeaterBoundUserInterface.cs:
Most other UI's are will not pop up when the device is not powered. This UI pops up, even when clicking on an unpowered space heater. This might make sense for a space heater, but it is different from the other machines. I wanted to make sure that you intended this.

Menshin replied:
I mimicked heaters/freezers UI behavior, but i can change that if it's not the expected behavior.

Partmedia commented on Resources/Prototypes/Entities/Structures/Piping/Atmospherics/portable.yml:
```diff
@@ -94,13 +96,86 @@
+
+- type: entity
+  id: SpaceHeater
+  parent: PortableAtmosMachine
```
I think this works for the portable scrubber because "portable" is in the name and the sprite has wheels, but I was scratching my head for a while until I realized that this machine doesn't start anchored.

Menshin replied:
I've always seen portable scrubbers, portable vents and space heaters as in the same category, but that's confusing indeed. Should i rename that to `unanchoredAtmosMachine` or something like that ?

Partmedia commented on Resources/Textures/Structures/Piping/Atmospherics/Portable/portable_sheater.rsi/sheaterCool.png:
I know this shows up in your video, but for some reason when I test this locally, I don't see the cooling or heating sprite animation?
![image](https://github.com/space-wizards/space-station-14/assets/3229565/a1ede645-b33b-4a43-ad9e-c9e8c1393f8b)

Partmedia reviewed (commented):
Unfortunately I was not able to get this to work in local testing. See video below:

Uploading out-2024-02-15_22.20.29.mp4…

Menshin commented:
> the stats seem weak to me, tile air generally needs **a lot** more heat to cool or heat than pipe air

What would be a good target for these ? It's actually as efficient as a heater (but less consuming) and far less effecient as a cooler (theorically).

Ilya246 commented:
> > the stats seem weak to me, tile air generally needs **a lot** more heat to cool or heat than pipe air
> 
> What would be a good target for these ? It's actually as efficient as a heater (but less consuming) and far less effecient as a cooler (theorically).

honestly i'm not even sure, if you were to give them very high efficiency you'd be able to break thermodynamics easily, and high power consumption means heavy power flicker and other potential issues
i would suggest efficiency between 2x and 3x and a 10 or 15 kW power consumption

Menshin commented:
I'll try these (seems a lot to me, will stuck to your lower bounds for now) and get back with the results. Thanks.

VektorZ1 commented:
Is there an improved version for RnD?

Menshin commented:
> Is there an improved version for RnD?

This PR just add regular space heaters, but it should be fairly easy to add an improved version for RnD when we finished balancing this one.

Menshin commented:
Should be ok for review now :

* redid the UI to account for the atmos equalization threshold (OP video updated)
* added a variant with enabled+anchored space heater for mappers
* added a popup when trying to use the space heater while it's not anchored
* made the space heater board researchable and printable (tied it to the tech unlocking the portable scrubber).
* cleaned up the code

Partmedia commented on Content.Client/Atmos/UI/SpaceHeaterBoundUserInterface.cs:
```diff
@@ -0,0 +1,83 @@
+using JetBrains.Annotations;
+using Robust.Client.UserInterface.Controls;
+
+namespace Content.Client.Atmos.UI
```
Please use file-scoped namespaces.

Partmedia commented on Content.Client/Atmos/UI/SpaceHeaterWindow.xaml.cs:
```diff
@@ -0,0 +1,59 @@
+using Content.Shared.Atmos.Piping.Portable.Components;
+using Content.Shared.Atmos;
+
+namespace Content.Client.Atmos.UI
```
Please use file-scoped namespaces.

Partmedia commented on Content.Client/Atmos/UI/SpaceHeaterWindow.xaml.cs:
```diff
@@ -0,0 +1,59 @@
+    [GenerateTypedNameReferences]
+    public sealed partial class SpaceHeaterWindow : DefaultWindow
+    {
+        //To account for a minimum delta temperature for atmos equalization to trigger we use a fixed step for target temperature increment/decrement
```
```suggestion
        // To account for a minimum delta temperature for atmos equalization to trigger we use a fixed step for target temperature increment/decrement.
```

Partmedia commented on Content.Client/Atmos/UI/SpaceHeaterWindow.xaml.cs:
```diff
@@ -0,0 +1,59 @@
+                Mode.AddItem(Loc.GetString($"comp-space-heater-mode-{value}"), (int) value);
+            }
+
+            //Only allow temperature increment/decrement of TemperatureChangeDelta
```
```suggestion
            // Only allow temperature increment/decrement of TemperatureChangeDelta
```

Partmedia commented on Content.Server/Atmos/Portable/SpaceHeaterComponent.cs:
```diff
@@ -0,0 +1,54 @@
+using Content.Shared.Atmos.Piping.Portable.Components;
+using Content.Shared.Atmos.Visuals;
+
+namespace Content.Server.Atmos.Portable
```
namespace

Partmedia commented on Content.Server/Atmos/Portable/SpaceHeaterComponent.cs:
```diff
@@ -0,0 +1,54 @@
+        ///     Coefficient of performance. Output power / input power.
+        ///     Positive for heaters, negative for freezers.
+        /// </summary>
+        [DataField("heatingCoefficientOfPerformance")]
+        [ViewVariables(VVAccess.ReadWrite)]
+        public float HeatingCp = 4f; // output power / input power
+
+        [DataField("coolingCoefficientOfPerformance")]
+        [ViewVariables(VVAccess.ReadWrite)]
+        public float CoolingCp = -3f; // output power / input power
```
I would prefer that heating $C_P$ be closer to 1. Cooling $C_P$ exceeds 1 for real heat pumps, but only because the cooling is being accomplished by moving heat elsewhere. Given that there is no cooling pipe, this should probably also be closer to 1. But I won't fight you on the cooling $C_P$ if you insist on this.

You may need to adjust the power consumption up to account for the change here.

Partmedia commented on Content.Server/Atmos/Portable/SpaceHeaterComponent.cs:
```diff
@@ -0,0 +1,54 @@
+        [ViewVariables(VVAccess.ReadWrite)]
+        public float CoolingCp = -3f; // output power / input power
+
+        /// <summary>
+        ///     Coefficient of performance. Output power / input power.
+        ///     Positive for heaters, negative for freezers.
+        /// </summary>
```
Copy/paste error in comments

Partmedia commented on Content.Server/Atmos/Portable/SpaceHeaterSystem.cs:
```diff
@@ -0,0 +1,172 @@
+using Content.Shared.UserInterface;
+using Robust.Server.GameObjects;
+
+namespace Content.Server.Atmos.Portable
```
namespace

Partmedia commented on Content.Server/Atmos/Portable/SpaceHeaterSystem.cs:
```diff
@@ -0,0 +1,172 @@
+                return;
+            }
+
+            //First get the heat direction of the thermomachine (if any) and update appeareance accordingly
```
```suggestion
            // First get the heat direction of the thermomachine (if any) and update appeareance accordingly
```

Partmedia commented on Content.Server/Atmos/Portable/SpaceHeaterSystem.cs:
```diff
@@ -0,0 +1,172 @@
+            }
+
+            //First get the heat direction of the thermomachine (if any) and update appeareance accordingly
+            if(thermoMachine.LastEnergyDelta > 0)
+            {
+                spaceHeater.State = SpaceHeaterState.Heating;
+            }
+            else if(thermoMachine.LastEnergyDelta < 0)
```
whitespace

Partmedia commented on Content.Server/Atmos/Portable/SpaceHeaterSystem.cs:
```diff
@@ -0,0 +1,172 @@
+
+            UpdateAppearance(uid, spaceHeater.State);
+
+            //Then, if in automatic temperature mode, check if we need to adjust the heat exchange direction
```
```suggestion
            // Then, if in automatic temperature mode, check if we need to adjust the heat exchange direction
```

Partmedia commented on Content.Shared/Atmos/Portable/Components/SharedSpaceHeaterComponent.cs:
```diff
@@ -0,0 +1,69 @@
+using Robust.Shared.Serialization;
+
+namespace Content.Shared.Atmos.Piping.Portable.Components
```
namespace

Partmedia commented on Resources/Locale/en-US/components/space-heater-component.ftl:
```diff
@@ -0,0 +1,13 @@
+﻿comp-space-heater-ui-thermostat = Thermostat :
```
```suggestion
﻿comp-space-heater-ui-thermostat = Thermostat:
```

Partmedia commented on Resources/Locale/en-US/components/space-heater-component.ftl:
```diff
@@ -0,0 +1,13 @@
+﻿comp-space-heater-ui-thermostat = Thermostat :
+comp-space-heater-ui-toggle = Toggle
```
I don't personally think a label for this is necessary, but it's fine if you want to keep it.

Partmedia commented on Resources/Locale/en-US/components/space-heater-component.ftl:
```diff
@@ -0,0 +1,13 @@
+comp-space-heater-mode-Heat = Heat
+comp-space-heater-mode-Cool = Cool
+
+comp-space-heater-unanchored = The device isn't anchored
```
```suggestion
comp-space-heater-unanchored = The device is not anchored.
```
Even better is if you can pass the `{$device}` in as a localized string and then change it to `THE{$device} is not anchored.` so that this can be used for other devices, too.

Partmedia commented on Resources/Locale/en-US/wires/wire-names.ftl:
```diff
@@ -38,6 +38,7 @@ wires-board-name-windoor = Windoor Control
 wires-board-name-mech = Mech
 wires-board-name-fatextractor = FatExtractor
 wires-board-name-flatpacker = Flatpacker
+wires-board-name-spaceheater = SpaceHeater
```
```suggestion
wires-board-name-spaceheater = Space Heater
```

Partmedia commented on Resources/Prototypes/Entities/Objects/Devices/Circuitboards/Machine/production.yml:
```diff
@@ -464,6 +464,22 @@
+  id: SpaceHeaterMachineCircuitBoard
+  parent: BaseMachineCircuitboard
+  name: space heater machine board
+  description: A PCB for a space heater.
```
Please expand the abbreviation "PCB". I can tell this was copy/pasted from scrubbers.

Menshin replied:
> I can tell this was copy/pasted from scrubbers.
It was 😁  (fixing both)

Partmedia commented on Resources/Prototypes/Entities/Structures/Piping/Atmospherics/portable.yml:
```diff
@@ -94,13 +90,103 @@
+            acts: [ "Destruction" ]
+
+- type: entity
+  id: SpaceHeater
```
Better to add a "[unanchored]" suffix here.

Partmedia commented on Resources/Prototypes/Entities/Structures/Piping/Atmospherics/portable.yml:
```diff
@@ -94,13 +90,103 @@
+- type: entity
+  parent: SpaceHeater
+  id: SpaceHeaterEnabled
+  suffix: Enabled
```
And turn this into an "[anchored]" version for mappers, but you don't need to enable it. Or have both "[anchored]" and "[enabled]" versions.

Partmedia reviewed (changes requested):
Thank you for your work on this. This is working much better for me in local testing. Now that we've gotten the big issues out of the way, I've left some nitpicks in the review comments. Please take a look when able. Thanks.

Menshin commented:
As requested, realigned coefficients of performance closer to 1 and cleaned the code.
To account for different usage scenarios, added a switch for the space heater power level that can halves or double the heating/cooling power (picture added to the PR).

If no other review requests arise, the PR should be done (we can always finetune the power consumption after playtesting, if needed). Thanks for all the feedback.

Partmedia commented on Content.Client/Atmos/UI/SpaceHeaterWindow.xaml.cs:
```diff
@@ -0,0 +1,71 @@
+{
+    // To account for a minimum delta temperature for atmos equalization to trigger we use a fixed step for target temperature increment/decrement
+    public int TemperatureChangeDelta = 5;
+    public bool Active = false;
+    public float MinTemp = 0.0f;
+    public float MaxTemp = 0.0f;
```
Comment with units

Partmedia replied:
These are all the defaults, and do not need to be explicitly specified, unless the defaults are non-obvious and deserve to be explicitly specified.

Partmedia commented on Content.Client/Atmos/UI/SpaceHeaterWindow.xaml.cs:
```diff
@@ -0,0 +1,71 @@
+
+    public void SetTemperature(float targetTemperature)
+    {
+        Thermostat.SetText((targetTemperature - Atmospherics.T0C).ToString() + "°C");
```
It would be best for this to be a FTL string.

Partmedia commented on Content.Server/Atmos/Portable/SpaceHeaterSystem.cs:
```diff
@@ -0,0 +1,198 @@
+        // First get the heat direction of the thermomachine (if any) and update appeareance accordingly
+        if (thermoMachine.LastEnergyDelta > 0)
+        {
+            spaceHeater.State = SpaceHeaterState.Heating;
```
`spaceHeater.State` seems to duplicate state that is stored elsewhere. This makes it confusing to understand and maintain, e.g. who is responsible for setting `spaceHeater.State` when something changes?

For example, it seems to me that `spaceHeater.State` is used only in `UpdateApperance()`. If that is the case, then this block computing what the state should be moved to `UpdateApperance()` and the extra calls to set it removed.

Menshin replied:
You're right, that's a leftover of an earlier attempt and the space heater state isn't used for anything besides the visualizer. It's gone now.

Partmedia commented on Content.Server/Atmos/Portable/SpaceHeaterSystem.cs:
```diff
@@ -0,0 +1,198 @@
+            return;
+
+        _power.TogglePower(uid);
+        if (powerReceiver.PowerDisabled)
+            spaceHeater.State = SpaceHeaterState.Off;
+        else
+            spaceHeater.State = SpaceHeaterState.StandBy;
+
```
For example, doesn't this duplicate a check that is made immediately in `UpdateAppearance()`?

Partmedia reviewed (commented):
A few nitpicks on conventions:
- Space between comments and the start of a comment, e.g. `// this` and not `//this`
- [No space between casts and what is being casted](https://learn.microsoft.com/en-us/dotnet/csharp/programming-guide/types/casting-and-type-conversions) `(int)this` and not `(int) this`

A few more nitpicks are in line. Thanks in advance for your work on this.

Menshin commented:
> A few nitpicks [...]

You call them nitpicks, i call them valid remarks on overall code quality.
All should (hopefully) be adressed now.

Partmedia reviewed (approved):
Thank you for your work on this.

Partmedia commented:
I've adjusted your Changelog entry per the [Changelog guidelines](https://docs.spacestation14.com/en/general-development/codebase-info/pull-request-guidelines.html#writing-an-effective-changelog).
